### PR TITLE
Vendor perseus plugin

### LINKED ISSUE #8246: Describe proper workflow for uploading and updating Perseus translations files
## Observed behavior
Extracting Perseus translatable strings is a murk-ish _affaire_, as the desired workflow is not completely obvious. We should document it properly.

## Errors and logs
…

## Expected behavior
…

## User-facing consequences
…

## Steps to reproduce
…

## Context
…

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary
* Creates a build pipeline for the main Perseus assets within the Kolibri repo
* Commits the results of this build pipeline into the repo
* Limits building of the Perseus renderer to just the Vue js components
* Removes dependency on external plugin

## References
Fixes #8246 by making the perseus plugin no different to any other in the repo

Includes some updates to the Learning Equality Perseus repo to make this easier, and to hardcode some of the changes we used to make at build time: https://github.com/learningequality/perseus/compare/f0a110a942d779d6172133a2c3750fdfd660e4be...learningequality:master

## Reviewer guidance
Do exercises still work?

----

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests

## PR process

- [ ] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

**Comment by indirectlylit** [comment]
> * Creates a build pipeline for the main Perseus assets within the Kolibri repo
> * Commits the results of this build pipeline into the repo

would it be straightforward to split these into two commits? If not that's okay but looking at 8a38cf5f3f306cbb2dbe828ed3d403d55768cdb9 it's hard to see just the new pipeline

**Comment by rtibbles** [comment]
> would it be straightforward to split these into two commits?

Yes, as long as you can abide a force push!

File: packages/kolibri-tools/lib/webpack.config.base.js [hunk 1]
```diff
@@ -144,20 +144,12 @@ module.exports = (data, { mode = 'development', hot = false } = {}) => {
...
           },
         },
-        // Use url loader to load font files.
-        {
-          test: /\.(eot|woff|ttf|woff2)$/,
-          use: {
-            loader: 'url-loader',
-            options: { name: '[name]-[contenthash].[ext]' },
-          },
-        },
```
Comment (indirectlylit) [inline packages/kolibri-tools/lib/webpack.config.base.js:160]: any risk of this introducing FOUT?

**Comment by indirectlylit** [comment]
sure thing

**Comment by radinamatic** [comment]
Perseus exercises work in both KA EN & KA ES resource. There are some `VueIntl` errors in the console, but only on exercises in English, and I think they might have been there even before this PR..

English | Spanish
--- | ---
LEDev2104 (start)  Running  - Oracle VM VirtualBox_001 | LEDev2104 (start)  Running  - Oracle VM VirtualBox_002

**Comment by rtibbles** [comment]
Yeah, this is actually a general issue with the most recent changes to the translation pipeline - the message files for English are not getting generated during development or production builds.

**Comment by radinamatic** [comment]
Gotcha, seems good to go then! 💯

**Review by radinamatic — APPROVED** [review]
LGTM! 👍🏽

File: kolibri/plugins/perseus_viewer/assets/dist/README.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+The files in this directory are generated using the buildPerseus script inside this plugin.
+They should be regenerated using the same command in the unlikely event that Perseus ever needs
+to be updated.
```
Comment (indirectlylit) [inline kolibri/plugins/perseus_viewer/assets/dist/README.md]: eot and woff2 files need to be tracked in git LFS:

```bash
git lfs track "*.eod"
git lfs track "*.woff2"
```

ref: https://github.com/learningequality/kolibri/blob/release-v0.15.x/.gitattributes

File: kolibri/plugins/perseus_viewer/buildPerseus.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+process.chdir(__dirname);
+
+const target = path.resolve(__dirname, './assets/dist');
+
+console.log('Clearing dist folder');
+
+rimraf.sync(target);
+
+mkdirp.sync(target);
```
Comment (indirectlylit) [inline kolibri/plugins/perseus_viewer/buildPerseus.js]: Trying to make this a bit easier to read

```suggestion
console.log('Cloning the Learning Equality Perseus repo...');
const submodulesPath = path.resolve(__dirname, './submodules');
rimraf.sync(submodulesPath);
mkdirp.sync(submodulesPath);
process.chdir(submodulesPath);
execSync('git clone https://github.com/learningequality/perseus.git');

console.log('Fetching all perseus submodules...');
process.chdir('perseus');
execSync('git submodule update --init --recursive');

console.log('Installing perseus dependencies...');
execSync('yarn install');

console.log('Clearing dist folder...');
process.chdir(__dirname);
const target = path.resolve(__dirname, './assets/dist');
rimraf.sync(target);
mkdirp.sync(target);
```

File: kolibri/plugins/perseus_viewer/buildPerseus.js [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  path.resolve(__dirname, './submodules/perseus/lib/katex/katex.css'),
+  path.resolve(__dirname, './submodules/perseus/build/perseus.css'),
+  path.resolve(__dirname, './submodules/perseus/lib/mathquill/mathquill.css'),
+];
+
+const compiler = webpack(buildConfig);
+
+console.log('Creating Perseus bundle with webpack');
+
+const tokenStackRegex = /_token_stack:\n/g;
```
Comment (indirectlylit) [inline kolibri/plugins/perseus_viewer/buildPerseus.js]: a comment would be helpful - what is this?

File: yarn.lock [hunk 5]
```diff
@@ -9593,7 +10588,7 @@ path-exists@^4.0.0:
   resolved "https://registry.yarnpkg.com/path-exists/-/path-exists-4.0.0.tgz#513bdbe2d3b95d7762e8c1137efa195c6c61b5b3"
   integrity sha512-ak9Qy5Q7jYb2Wwcey5Fpvg2KoAc/ZIhLSLOSBmRmygPsGwkVVt0fZa0qrtMz+m6tJTAHfZQ8FnmB4MG4LWy7/w==
 
-path-is-absolute@^1.0.0:
+path-is-absolute@^1.0.0, path-is-absolute@^1.0.1:
```
Comment (indirectlylit) [inline yarn.lock:10484]: A number of dependencies (such as this one) are now duplicated.

Maybe worth filing a follow-up tech debt issue to find the causes and, where possible, try to consolidate on the most recent version?

File: kolibri/plugins/perseus_viewer/buildPerseus.js [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+const README = `The files in this directory are generated using the buildPerseus script inside this plugin.
+They should be regenerated using the same command in the unlikely event that Perseus ever needs
+to be updated.`;
```
Comment (indirectlylit) [inline kolibri/plugins/perseus_viewer/buildPerseus.js]: Maybe provide a link here to the section of the developer docs which describe how to do this? ;-)

But more seriously, the why what and how should get some documentation either in comments in this file or a mention in the dev docs.

File: kolibri/plugins/perseus_viewer/buildPerseus.js [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+compiler.run(err => {
```
Comment (indirectlylit) [inline kolibri/plugins/perseus_viewer/buildPerseus.js:71]: this block could some comments indicating what's about to happen

File: kolibri/plugins/perseus_viewer/assets/src/translator.js [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  'a *simplified improper* fraction, like $7/4$': 'a *simplified improper* fraction, like $7/4$',
+  'a mixed number, like $1 3/4$': 'a mixed number, like $1\\\\ 3/4$',
+  'an *exact* decimal, like $0.75$': 'an *exact* decimal, like $0.75$',
+  'a percent, like $12.34%$': 'a percent, like $12.34\\\\%$',
+  'a multiple of pi, like $12 text{pi}$ or $2/3 text{pi}$':
+    'a multiple of pi, like $12\\\\ \\\\text{pi}$ or $2/3\\\\ \\\\text{pi}$',
+  '**Your answer should be** ': '**Your answer should be** ',
+  'Your answer:': 'Your answer:',
+  '0 solutions': '0 solutions',
+  'Finite solutions': 'Finite solutions',
```
Comment (indirectlylit) [inline kolibri/plugins/perseus_viewer/assets/src/translator.js:20]: When running the build script, my version of this file got prettified which introduced an undesirable diff:


```diff
-  Hints: 'Hints',
-  'Get another hint': 'Get another hint',
-  'Hint #{ pos }': 'Hint #{ pos }',
-  'an integer, like $6$': 'an integer, like $6$',
-  'a *proper* fraction, like $1/2$ or $6/10$': 'a *proper* fraction, like $1/2$ or $6/10$',
-  'a *simplified proper* fraction, like $3/5$': 'a *simplified proper* fraction, like $3/5$',
-  'an *improper* fraction, like $10/7$ or $14/8$': 'an *improper* fraction, like $10/7$ or $14/8$',
-  'a *simplified improper* fraction, like $7/4$': 'a *simplified improper* fraction, like $7/4$',
-  'a mixed number, like $1 3/4$': 'a mixed number, like $1\\\\ 3/4$',
-  'an *exact* decimal, like $0.75$': 'an *exact* decimal, like $0.75$',
-  'a percent, like $12.34%$': 'a percent, like $12.34\\\\%$',
-  'a multiple of pi, like $12 text{pi}$ or $2/3 text{pi}$':
-    'a multiple of pi, like $12\\\\ \\\\text{pi}$ or $2/3\\\\ \\\\text{pi}$',
-  '**Your answer should be** ': '**Your answer should be** ',
-  'Your answer:': 'Your answer:',
-  '0 solutions': '0 solutions',
-  'Finite solutions': 'Finite solutions',
+  "Hints": "Hints",
+  "Get another hint": "Get another hint",
+  "Hint #{ pos }": "Hint #{ pos }",
+  "an integer, like $6$": "an integer, like $6$",
+  "a *proper* fraction, like $1/2$ or $6/10$": "a *proper* fraction, like $1/2$ or $6/10$",
+  "a *simplified proper* fraction, like $3/5$": "a *simplified proper* fraction, like $3/5$",
+  "an *improper* fraction, like $10/7$ or $14/8$": "an *improper* fraction, like $10/7$ or $14/8$",
+  "a *simplified improper* fraction, like $7/4$": "a *simplified improper* fraction, like $7/4$",
+  "a mixed number, like $1 3/4$": "a mixed number, like $1\\\\ 3/4$",
+  "an *exact* decimal, like $0.75$": "an *exact* decimal, like $0.75$",
+  "a percent, like $12.34%$": "a percent, like $12.34\\\\%$",
+  "a multiple of pi, like $12 text{pi}$ or $2/3 text{pi}$": "a multiple of pi, like $12\\\\ \\\\text{pi}$ or $2/3\\\\ \\\\text{pi}$",
+  "**Your answer should be** ": "**Your answer should be** ",
+  "Your answer:": "Your answer:",
+  "0 solutions": "0 solutions",
+  "Finite solutions": "Finite solutions",
```

File: kolibri/plugins/perseus_viewer/perseusBuildConfig.js [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+
+const node_modules = path.resolve(__dirname, './node_modules');
+
+const base_node_modules = path.resolve(__dirname, '../../../node_modules');
+
+module.exports = {
+  entry: path.resolve(__dirname, './submodules/perseus/src/perseus.js'),
+  output: {
+    path: path.resolve(__dirname, './assets/dist'),
+    filename: 'perseus.js',
```
Comment (indirectlylit) [inline kolibri/plugins/perseus_viewer/perseusBuildConfig.js:20]: This file contains references to the developer machine, which will lead to undesirable diffs when run on a new machine:

```diff
- // EXTERNAL MODULE: /home/richard/github/kolibri/node_modules/inline-style-prefixer/static/plugins/filter.js
+ // EXTERNAL MODULE: /Users/d/Projects/le/kolibri/node_modules/inline-style-prefixer/static/plugins/filter.js
```


aside: kind of cool that our `./kolibri` directories are exactly the same number of characters into their respective absolute paths...

File: docs/i18n.rst [hunk 10]
```diff
@@ -248,8 +237,6 @@ Typically, strings will be uploaded when a new release branch is cut from ``deve
 
 Before translators can begin working on the strings in our application, they need to be uploaded to Crowdin. Translations are maintained in release branches on Crowdin in the `Crowdin kolibri project <http://crowdin.com/project/kolibri>`__.
 
-.. warning:: You *must* have the Perseus plugin installed for development as described above, or not all strings will be extracted!
-
```
Comment (indirectlylit) [inline docs/i18n.rst:252]: Could add a note about updating and rebuilding the perseus repo to the frontend build pipeline docs:

https://kolibri-dev.readthedocs.io/en/develop/frontend_architecture/frontend_build_pipeline.html

**Review by indirectlylit — DISMISSED** [review]
Exciting to get this vendored at last!

Main changes requested:

* Fonts tracked in LFS
* Some additional docs and comments on how to update Perseus

Possible follow-up issues:

* Make sure diffs are cleaner after an update
* Avoid adding so many extra dependencies

File: packages/kolibri-tools/lib/webpack.config.base.js (hunk 1 above)
Comment (rtibbles) [reply packages/kolibri-tools/lib/webpack.config.base.js:160]: Possibly, but only within Perseus - as nowhere else in the Kolibri codebase were we leveraging this.

The issue is that Perseus has so many font files that inlining them all makes the CSS enormous.

File: kolibri/plugins/perseus_viewer/assets/dist/README.md (hunk 2 above)
Comment (rtibbles) [reply kolibri/plugins/perseus_viewer/assets/dist/README.md]: Ah, I thought git LFS had picked them up already, but I guess not.

File: kolibri/plugins/perseus_viewer/buildPerseus.js (hunk 4 above)
Comment (rtibbles) [reply kolibri/plugins/perseus_viewer/buildPerseus.js]: Oh - I can actually commit this change in the Perseus repo now I think about it.

It's a weird non-strict mode valid function label in the `KAS` jison-autogenerated parser module. It breaks building.

File: kolibri/plugins/perseus_viewer/assets/src/translator.js (hunk 8 above)
Comment (rtibbles) [reply kolibri/plugins/perseus_viewer/assets/src/translator.js:20]: I can lint the file before writing it using kolibri-tools, will update.

File: kolibri/plugins/perseus_viewer/perseusBuildConfig.js (hunk 9 above)
Comment (rtibbles) [reply kolibri/plugins/perseus_viewer/perseusBuildConfig.js:20]: Oh weird - that's not helpful at all.

File: kolibri/plugins/perseus_viewer/assets/dist/README.md (hunk 2 above)
Comment (rtibbles) [reply kolibri/plugins/perseus_viewer/assets/dist/README.md]: To remind myself, I am going to update this to only include `.woff` files, exclusively, and update the CSS accordingly - we determined that `.woff` is maximally compatible for the platforms we support.

File: yarn.lock (hunk 5 above)
Comment (rtibbles) [reply yarn.lock:10484]: Filed here: https://github.com/learningequality/kolibri/issues/8354

**Review by indirectlylit — APPROVED** [review]
thanks for addressing my feedback! good to go in terms of code review – assuming testing checks out, let's get this in

**Comment by pcenov** [comment]
Hi @rtibbles I attempted to have a look at this with the following [build](https://buildkite.com/learningequality/kolibri-python-package/builds/3497) and for some reason after successful installation both Windows and Debian builds are showing a 500 Server Error instead of the setup wizard:


Attaching the logs as well:
[WindowsLogs.zip](https://github.com/learningequality/kolibri/files/7065054/WindowsLogs.zip)

File: yarn.lock [hunk 11]
```diff
@@ -11267,7 +12407,7 @@ sax@^1.2.4, sax@~1.2.4:
   resolved "https://registry.yarnpkg.com/sax/-/sax-1.2.4.tgz#2816234e2378bddc4e5354fab5caa895df7100d9"
   integrity sha512-NqVDv9TpANUjFm0N8uM5GxL36UgKi9/atZw+x7YFnQ8ckwFGKrl4xX4yWtrey3UJm5nP1kUbnYgLopqWNSRhWw==
 
-schema-utils@^0.4.0:
+schema-utils@^0.4.0, schema-utils@^0.4.4, schema-utils@^0.4.5:
```
Comment (indirectlylit) [inline yarn.lock]: example of new patch versions being added

**Review by indirectlylit — DISMISSED** [review]
> Fixes #8246 by making the perseus plugin no different to any other in the repo

Something may have regressed here? Looking at _yarn.lock_ I see lots of new patch versions added and very few removed

EDIT: please ignore – I was looking at the wrong issue

File: yarn.lock (hunk 11 above)
Comment (indirectlylit) [reply yarn.lock]: oh my fault! I thought this fixed #8354 but it was actually just opened here

**Comment by radinamatic** [comment]
Looking good in Windows 7 on all three browsers (FF; Chrome & IE11)! 👍🏽

**Comment by radinamatic** [comment]
Exercises in the most recent build working without issues in Firefox, Chrome and Edge on Windows 10! 💯 
I also used an 8.1.0 Android device as a client, and seen happy ✔️  in mobile Firefox and Chrome 🙂 

@pcenov Please run this on your Mac mini on Monday to confirm that Safari is also on board with all the changes, thank you!

**Comment by pcenov** [comment]
Tested with the following [build](https://buildkite.com/learningequality/kolibri-macos/builds/3832) and didn't observe any issues with the perseus exercises in both Safari and the MacOS app:

**Comment by indirectlylit** [comment]
awesome! sounds ready to merge?

**Pull request merged**